### ISSUE TITLE
False positive `protected-access` when accessing a different instance of the same class

### ISSUE THREAD

**Issue by macdjord** [PyCQA/pylint#8450]
### Bug description

`test.py`:
```python
"""Test"""

import dataclasses as _dataclasses
import typing as _tp

@_dataclasses.dataclass(slots=True, frozen=True)
class RecursiveClass:
    value: str
    child: _tp.Optional["RecursiveClass"]

    def do_something(self) -> list:
        return self._do_something()

    def _do_something(self) -> list:
        if self.child is None:
            return [self.value]
        return [self.value, self.child._do_something()]


r0 = RecursiveClass("a", None)
r1 = RecursiveClass("b", r0)
```

### Configuration

```ini
[MASTER]
disable=
    logging-fstring-interpolation,
    f-string-without-interpolation,
    global-statement,
    fixme,
    too-many-branches,
    too-many-arguments,
    too-few-public-methods,
    too-many-ancestors,
    missing-class-docstring,
    missing-function-docstring,

[FORMAT]
max-line-length = 120
expected-line-ending-format = LF
indent-string = "    "


[BASIC]
good-names = i,j,k,n,ex,fd,__
include-naming-hint = true
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:17:28: W0212: Access to a protected member _do_something of a client class (protected-access)

------------------------------------------------------------------
Your code has been rated at 9.23/10 (previous run: 9.23/10, +0.00)
```

### Expected behavior

Member `child` of class `RecursiveClass` is itself an instance of `RecursiveClass` (or `None`), as is both documented via the type hints and demonstrated via line 21 (where an instance of `RecursiveClass`is created with another instance as its `child` value). Thus no warning should be emitted about access a protected member of it.

### Pylint version

```shell
pylint 2.16.2 
astroid 2.14.2
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

Window 10

**Comment by davetapley** [comment]
This also occurs when using `typing.Self` e.g.:

```py
@dataclass
class Foo:
    _x: int
    def merge(self, other: Self):
            self._x += other._x
```

**Comment by Pierre-Sassoulas** [comment]
I added the "need decision" label because I think this could be working as intended, the other object is external to the instance even if it's the same class.

**Comment by DanielNoord** [comment]
Yeah, I would say this is indeed working as intended. Instances of the same class are not internal to an instance.
